### ISSUE TITLE
Bugs reported from @pjakubczyk

### ISSUE THREAD

**Issue by Centril** [Centril/gradle-plugin-robospock#6]
- Ensure JDK1.7 compatibility.
- Replace android plugin references to strings in `RoboSpockConfiguration.isAndroid`:
- Ensure that it works for libraries - don't rely on applicationVariant.

**Comment by pjakubczyk** [comment]
Let's continue the conversation here :)

**Comment by Centril** [comment]
=)

I very much appreciate pull-requests for anything, as I'm taking 3 courses at [Chalmers University of Technology](http://en.wikipedia.org/wiki/Chalmers_University_of_Technology) at the moment, so the time I have is unfortunately limited ^^

**Comment by Centril** [comment]
This issue should be fixed now - does it work for you?

**Comment by pjakubczyk** [comment]
Java 1.7 works!

**Comment by pjakubczyk** [comment]
but I'm having Caused by: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object '21.1.0' with class 'com.android.sdklib.repository.FullRevision' to class 'com.android.sdklib.repository.FullRevision'
    at com.jakewharton.sdkmanager.internal.PackageResolver.resolveBuildTools(PackageResolver.groovy:66)
    at com.jakewharton.sdkmanager.internal.PackageResolver.resolve(PackageResolver.groovy:58)
    at com.jakewharton.sdkmanager.internal.PackageResolver$resolve$0.call(Unknown Source)
    at com.jakewharton.sdkmanager.internal.PackageResolver.resolve(PackageResolver.groovy:20)
    at com.jakewharton.sdkmanager.internal.PackageResolver$resolve.call(Unknown Source)
    at se.centril.robospock.RoboSpockAction.fixSupportLib(RoboSpockAction.groovy:87)
    at se.centril.robospock.RoboSpockAction$_execute_closure2.doCall(RoboSpockAction.groovy:69)
    at se.centril.robospock.RoboSpockAction.execute(RoboSpockAction.groovy:69)
    at se.centril.robospock.RoboSpockAction$execute$1.call(Unknown Source)
    at se.centril.robospock.RoboSpockAction$_perform_closure1.doCall(RoboSpockAction.groovy:58)

**Comment by Centril** [comment]
That is super super strange... It is casting an object of class A, to class A - which should always work... Do you have a repository with the code you are testing with so that I can reproduce it?

**Comment by bluesliverx** [comment]
I worked around the FullRevision problem by assigning it to a "def" variable instead of "FullRevision".  I overrode it on the metaClass before applying the plugin:

``` groovy

com.jakewharton.sdkmanager.internal.PackageResolver.metaClass.resolveBuildTools { ->
    def buildToolsRevision = project.android.buildToolsRevision
    log.debug "Build tools version: $buildToolsRevision"

    def buildToolsRevisionDir = new File(buildToolsDir, buildToolsRevision.toString())
    if (folderExists(buildToolsRevisionDir)) {
        log.debug 'Build tools found!'
        return
    }

    log.lifecycle "Build tools $buildToolsRevision missing. Downloading..."

    def code = androidCommand.update "build-tools-$buildToolsRevision"
    if (code != 0) {
        throw new StopExecutionException("Build tools download failed with code $code.")
    }
}
```

My guess is that the classloaders for the FullRevision classes are different, hence java thinks they are two different classes.   I'm not sure when the different class loaders are coming in, but that's my guess.  Anyway, changing it to a "def" fixes the issue for now.  By the way, I'm using Java 7 so apparently those fixes for JDK7 worked.

I'm now working through a new problem from gradle:

```
Could not find com.android.support:support-v4:19.0.1
```

**Comment by Centril** [comment]
Thanks @bluesliverx - I totally forgot about classloaders ^^
I'll add it as a temporary fix for now... Please make a pull-request in the repo for `sdk-manager-plugin` as well so that everyone gets the fix and so that it doesn't have to be included in future versions.

Also, feel free to make pull-requests so that I can merge upstream faster =)

**Comment by Centril** [comment]
@pjakubczyk I just released v0.3.1 with the patch that @bluesliverx added & a cache speedup of the tests that should be helpful. Please verify that it now works for you.

**Comment by bluesliverx** [comment]
Submitted PR - https://github.com/JakeWharton/sdk-manager-plugin/pull/58.

**Comment by bluesliverx** [comment]
0.3.1 gives me a new error:

```
Caused by: org.gradle.api.internal.MissingMethodException: Could not find method sourceDir() for arguments [source set m
ain] on se.centril.robospock.RoboSpockConfiguration_Decorated@c6dd37.
        at org.gradle.api.internal.AbstractDynamicObject.methodMissingException(AbstractDynamicObject.java:68)
        at org.gradle.api.internal.AbstractDynamicObject.invokeMethod(AbstractDynamicObject.java:56)
        at org.gradle.api.internal.CompositeDynamicObject.invokeMethod(CompositeDynamicObject.java:172)
        at se.centril.robospock.RoboSpockConfiguration_Decorated.invokeMethod(Unknown Source)
        at se.centril.robospock.RoboSpockConfiguration.mainSourceDir(RoboSpockConfiguration.groovy:229)
        at se.centril.robospock.RoboSpockConfiguration$mainSourceDir$1.call(Unknown Source)
        at se.centril.robospock.RoboSpockTest.configure(RoboSpockTest.groovy:69)
        at se.centril.robospock.RoboSpockAction$_setupTestTask_closure7.doCall(RoboSpockAction.groovy:150)
        at org.gradle.api.internal.ClosureBackedAction.execute(ClosureBackedAction.java:59)
        at org.gradle.util.ConfigureUtil.configure(ConfigureUtil.java:130)
        at org.gradle.util.ConfigureUtil.configure(ConfigureUtil.java:110)
        at org.gradle.api.internal.AbstractTask.configure(AbstractTask.java:482)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.create(DefaultTaskContainer.java:79)
        at org.gradle.api.tasks.TaskContainer$create$1.call(Unknown Source)
        at se.centril.robospock.RoboSpockAction.setupTestTask(RoboSpockAction.groovy:148)
        at se.centril.robospock.RoboSpockAction$_execute_closure2.doCall(RoboSpockAction.groovy:70)
        at se.centril.robospock.RoboSpockAction.execute(RoboSpockAction.groovy:70)
        at se.centril.robospock.RoboSpockAction$execute$1.call(Unknown Source)
        at se.centril.robospock.RoboSpockAction$_perform_closure1.doCall(RoboSpockAction.groovy:59)
        at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java
:40)
        at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java
:25)
        at org.gradle.listener.BroadcastDispatch.dispatch(BroadcastDispatch.java:83)
        at org.gradle.listener.BroadcastDispatch.dispatch(BroadcastDispatch.java:31)
        at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.j
ava:93)
        at com.sun.proxy.$Proxy13.afterEvaluate(Unknown Source)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.notifyAfterEvaluate(LifecycleProjectEvaluator.java
:67)
        ... 30 more
```

**Comment by Centril** [comment]
Stacktrace and usage?

**Comment by bluesliverx** [comment]
I edited with the stacktrace, sorry.  The usage was just running "gradlew app-test:test".

**Comment by Centril** [comment]
Weird... It's trying to call a method in the same class that is there in the source code...
Don't get why you guys are getting all this weird bugs - it runs fine for me.

**Comment by bluesliverx** [comment]
It's not quite a simple project, but it's not overly complex either.  A couple of things I'm doing that you may not be:
- I have two library projects side-by-side with the project under test
- I downloaded all JARs into the `app/libs` directory (so I can develop on a table that doesn't support gradle)

**Comment by Centril** [comment]
I think I tracked down the error, it seems to be due to the fact that the robospock plugin and the android plugin are not applied to the classpath together, but separately...
The method `private File sourceDir( AndroidSourceSet ass )` is not recognized because AndroidSourceSet is from the android gradle plugin...

If you add both plugins (android & robospock) to the buildscript of the root project, it works for me, if i separate them, it breaks with the error you specified.

I'm working on a fix right now.

**Comment by bluesliverx** [comment]
Awesome, sounds good.  It also looks like that somewhat works when putting it on the tested project.  I still have problems with Android Studio not liking the tested project when it's done that way ("Cannot increment past a total of" error mentioned on #5).

I still am running into the issue about not being able to find the support library.  This may be something in the newest android gradle plugin (0.14.+). I may just pull down the plugin source into buildSrc and see what I can figure out.

**Comment by bluesliverx** [comment]
Created a new issue and pull request for the support library issue at #10 and #11.

**Comment by Centril** [comment]
Merged both PRs. I'm going to remove both the dependency on the android plugin & the sdk manager plugin and instead make use of the PackageResolver alone. I'll release a bugfix version tonight or tomorrow.

**Comment by pjakubczyk** [comment]
I tested the 0.3.1 and I had this problem
Could not find method sourceDir() for arguments [source set main] on se.centril.robospock.RoboSpockConfiguration_Decorated@35a1136e.

so the same as you guys :)

**Comment by bluesliverx** [comment]
Interestingly enough, I forked the plugin, pulled the code into buildSrc (just the plugin directory basically), made the patches I submitted as PRs yesterday, and it seems to be working now by applying the plugin to a separate test project.  Applying it to the same project without a tester project runs into the intelliJ/AS issue I mentioned in #5.

So apparently we're on the right track, or buildSrc pulls in dependencies with different class loading techniques so that the sourceDir issue is not a problem for me anymore...

**Comment by Centril** [comment]
So I finally released 0.3.2...

The [test-robospock-plugin](https://github.com/Centril/test-robospock-plugin) should outline exactly how to get the plugin working. The only change I've made is to require that the `android-sdk-manager` be added as a buildscript dependency on the project that the plugin is applied on... which can be done with:

``` groovy
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'com.jakewharton.sdkmanager:gradle-plugin:0.12.+'
        classpath 'se.centril.robospock:gradle-plugin-robospock:0.3.2'
    }
}
```

I really hope we don't get more weird bugs now...

Anyone who can test [test-robospock-plugin](https://github.com/Centril/test-robospock-plugin) and other stuff?

PS: I apologize for the wait, I've really been swamped lately with courses :/ If anyone wants to get collaborator status & help me out more, please say so.
